Targets/STM32F4/STM32F4_CAN.cpp

```diff
@@ -0,0 +1,1597 @@
+                          This parameter can be a value between 0 to 0x1FFFFFFF. */
+
+    uint8_t IDE;     /*!< Specifies the type of identifier for the message that
+                          will be received. This parameter can be a value of
+                          @ref CAN_identifier_type */
+
+    uint8_t RTR;     /*!< Specifies the type of frame for the received message.
+                          This parameter can be a value of
+                          @ref CAN_remote_transmission_request */
+
+    uint8_t DLC;     /*!< Specifies the length of the frame that will be received.
+                          This parameter can be a value between 0 to 8 */
+
+    uint8_t Data[8]; /*!< Contains the data to be received. It ranges from 0 to
+                          0xFF. */
+
+    uint8_t FMI;     /*!< Specifies the index of the filter the message stored in
+                          the mailbox passes through. This parameter can be a
+                          value between 0 to 0xFF */
+} STM32F4_Can_RxMessage;
+
+
+struct STM32F4_CanData_T {
+    uint32_t *matchFilters;
+    uint32_t matchFiltersSize;
+
+    uint32_t *lowerBoundFilters;
+    uint32_t *upperBoundFilters;
+    uint32_t groupFiltersSize;
+
+}canData[2];
+
+
+typedef struct {
+    uint32_t TimeStampL;
+    uint32_t TimeStampH;
+
+    uint32_t Frame; // Bits 16..19: DLC - Data Length Counter
+                    // Bit 30: Set if this is a RTR message
+                    // Bit 31: Set if this is a 29-bit ID message
+    uint32_t MsgID;	// CAN Message ID (11-bit or 29-bit)
+    uint32_t DataA;	// CAN Message Data Bytes 0-3
+    uint32_t DataB;	// CAN Message Data Bytes 4-7
+
+} STM32F4_Can_Message;
+
+enum class STM32F4_Can_Error : uint32_t {
+    OverRun,
+    RxOver,
+    BusOff,
+    ErrorPassive,
+    dataReveived = 0xFF,
+};
+
+struct STM32F4_Can_Controller {
+    const TinyCLR_Can_Provider* provider;
+
+    STM32F4_Can_Message *canRxMessagesFifo;
+
+    STM32F4_Can_TxMessage *txMessage;
+    STM32F4_Can_RxMessage *rxMessage;
+
+    STM32F4_Can_InitTypeDef initTypeDef;
+    STM32F4_Can_FilterInitTypeDef filterInitTypeDef;
+
+    TinyCLR_Can_ErrorReceivedHandler   errorEventHandler;
+    TinyCLR_Can_MessageReceivedHandler    messageReceivedEventHandler;
+
+    int32_t can_rx_count;
+    int32_t can_rx_in;
+    int32_t can_rx_out;
+
+    int32_t can_max_messages_receiving;
+
+    uint32_t baudrate;
+
+};
+
+static const STM32F4_Gpio_Pin g_STM32F4_Can_Tx_Pins[] = STM32F4_CAN_TX_PINS;
+static const STM32F4_Gpio_Pin g_STM32F4_Can_Rx_Pins[] = STM32F4_CAN_RX_PINS;
+
+static const int TOTAL_CAN_CONTROLLERS = SIZEOF_ARRAY(g_STM32F4_Can_Tx_Pins);
+
+static STM32F4_Can_Controller canController[TOTAL_CAN_CONTROLLERS];
+
+static TinyCLR_Can_Provider *canProvider[TOTAL_CAN_CONTROLLERS];
+static TinyCLR_Api_Info canApi;
+
+static uint8_t canProviderDefs[TOTAL_CAN_CONTROLLERS * sizeof(TinyCLR_Can_Provider)];
+
+void InsertionSort(uint32_t *arr, int32_t length) {
```

Comment:
See if InsertionSort and the other three sort methods below exist in the C++ STL. If they don't and we can't change our code to conform to what does, move these functions to a new top level "Shared" folder. Then make a separate PR to add that folder to the compile script.